From f2557ab57285c1d6aedcbe5f409ce34b23e5a0a9 Mon Sep 17 00:00:00 2001
From: Lorenzo Bianconi <lorenzo.bianconi@redhat.com>
Date: Thu, 14 May 2026 18:46:01 +0200
Subject: [PATCH] northd: Use MC_UNKNOWN for broadcast ARP requests

Broadcast ARP requests and ND_NS flooded to all non-router ports can exceed
the OVS resubmit limit in large logical switches. Send them to MC_UNKNOWN
instead; preserve MC_FLOOD_L2 for GARP.

Backport of ovn-org/ovn@f2557ab57285c1d6aedcbe5f409ce34b23e5a0a9
adapted to the OVN 24.03 commit pinned by Kube-OVN release-1.14.
Co-authored-by: Dumitru Ceara <dceara@redhat.com>
Signed-off-by: Dumitru Ceara <dceara@redhat.com>
Signed-off-by: Lorenzo Bianconi <lorenzo.bianconi@redhat.com>
Signed-off-by: Ales Musil <amusil@redhat.com>
---
 northd/northd.c | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/northd/northd.c b/northd/northd.c
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -8428,4 +8428,5 @@ build_lswitch_rport_arp_req_flow(
 {
     struct ds match   = DS_EMPTY_INITIALIZER;
     struct ds m       = DS_EMPTY_INITIALIZER;
+    struct ds m_garp  = DS_EMPTY_INITIALIZER;
     struct ds actions = DS_EMPTY_INITIALIZER;
@@ -8452,14 +8453,25 @@ build_lswitch_rport_arp_req_flow(
+    if (addr_family == AF_INET) {
+        ds_clone(&m_garp, &m);
+        ds_put_format(&m_garp, " && arp.spa == %s", ips);
+    }
+
     /* Send a the packet to the router pipeline.  If the switch has non-router
      * ports then flood it there as well.
      */
     if (od->n_router_ports != od->nbs->n_ports) {
         ds_put_format(&actions, "clone {outport = %s; output; }; "
-                                "outport = \""MC_FLOOD_L2"\"; output;",
+                                "outport = \""MC_UNKNOWN"\"; output;",
                       patch_op->json_key);
         ovn_lflow_add_with_hint(lflows, od, S_SWITCH_IN_L2_LKUP,
                                 priority, ds_cstr(&match),
                                 ds_cstr(&actions), stage_hint,
                                 lflow_ref);
+        if (addr_family == AF_INET) {
+            ovn_lflow_add_with_hint(lflows, od, S_SWITCH_IN_L2_LKUP,
+                                    priority + 10, ds_cstr(&m_garp),
+                                    "outport = \""MC_FLOOD_L2"\"; output;",
+                                    stage_hint, lflow_ref);
+        }
     } else {
         ds_put_format(&actions, "outport = %s; output;", patch_op->json_key);
         ovn_lflow_add_with_hint(lflows, od, S_SWITCH_IN_L2_LKUP,
@@ -8476,7 +8488,7 @@ build_lswitch_rport_arp_req_flow(
         ds_clear(&actions);
         if (od->n_router_ports != od->nbs->n_ports) {
             ds_put_format(&actions, "clone {outport = %s; output; }; "
-                                    "outport = \""MC_FLOOD_L2"\"; output;",
+                                    "outport = \""MC_UNKNOWN"\"; output;",
                           patch_op->cr_port->json_key);
             ovn_lflow_add_with_hint(lflows, od, S_SWITCH_IN_L2_LKUP,
                                     priority, ds_cstr(&match),
@@ -8495,4 +8507,5 @@ build_lswitch_rport_arp_req_flow(
     ds_destroy(&m);
     ds_destroy(&match);
+    ds_destroy(&m_garp);
     ds_destroy(&actions);
 }
@@ -9714,6 +9727,6 @@ build_lswitch_destination_lookup_bmcast(struct ovn_datapath *od,
                        "broadcast-arps-to-all-routers", true)) {
         ovn_lflow_add(lflows, od, S_SWITCH_IN_L2_LKUP, 72,
                       "eth.mcast && (arp.op == 1 || nd_ns)",
-                      "outport = \""MC_FLOOD_L2"\"; output;",
+                      "outport = \""MC_UNKNOWN"\"; output;",
                       lflow_ref);
     }
